### ISSUE TITLE
Feature: move all tasks to event-based output

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#200]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
 - API
 - UI

##### SUMMARY

AWX currently displays job runs in an events-based manner - the stdout is reconstructed from the actual Ansible job events from the callback.

This should be done for all tasks:
- [ ] ad-hoc commands
- [ ] inventory updates
- [ ] project updates
- [ ] system jobs

ad-hoc commands already have events, the UI just needs to be based on them. Project updates will need to be migrated to events, and the UI updated. Inventory updates and system jobs need to somehow generate events, and then have the UI updated.

**Comment by wenottingham** [comment]
As part of this, the stdout for these tasks will likely no longer be stored in the database directly.

**Comment by ryanpetrello** [comment]
Probably system jobs, too.

**Comment by wenottingham** [comment]
Added system jobs to initial spec above.

**Comment by jamm1985** [comment]
I have the same issue with ad-hoc commands and `stdout capture is missing` #146 

![image](https://user-images.githubusercontent.com/3518847/30680574-15c97046-9eec-11e7-9161-da7201ae806a.png)

```
CONTAINER ID        IMAGE                COMMAND                  CREATED             STATUS              PORTS                                NAMES
7ba282a1088d        awx_task:1.0.0.505   "/tini -- /bin/sh ..."   23 hours ago        Up 23 hours         8052/tcp                             awx_task
3e7531621947        awx_web:1.0.0.505    "/tini -- /bin/sh ..."   23 hours ago        Up 23 hours         0.0.0.0:80->8052/tcp                 awx_web
05e6977d0f6c        memcached:alpine     "docker-entrypoint..."   23 hours ago        Up 23 hours         11211/tcp                            memcached
a4e24611607e        rabbitmq:3           "docker-entrypoint..."   23 hours ago        Up 23 hours         4369/tcp, 5671-5672/tcp, 25672/tcp   rabbitmq
```

**Comment by mirakl577** [comment]
Hello, is there any workaround until its not resolved ?

can't see the progress, when job is running, hard to understand what is going on

**Comment by ryanpetrello** [comment]
One consideration for this change is that job output gathered _prior_ to this feature will still be stdout blobs stored in the database - we need a migration strategy for these, and bursting them out to job events *after* the fact is probably too complicated (and time consuming) of a migration.

One solution we've thrown around is the idea of *moving* this data into a separate deprecated/legacy table away from `main_unifiedjob` that the UnifiedJob table is smart enough to load on-demand as needed (i.e., for stdout API endpoints).  This would also alleviate performance issues we've seen related to multi-MB stdout blobs affecting general query performance on `main_unifiedjob`.

**Comment by ryanpetrello** [comment]
Problems to Solve
=============

   - Storing stdout on local file systems isn't reliable in clustered environments (see https://github.com/ansible/awx/issues/146)
    - Storing large stdout blobs on the `main_unifiedjob` table can result in poor query performance for certain ORM-generated queries (see https://github.com/ansible/awx/commit/1bb6c17fe27d0deacc3bd0a3b763903fa35725df)
    - The UI (and other API consumers) should have *one* unified way to represent job output (keep websockets for real time/streaming and API for historical event data; stop UI usage of the `/api/v2/jobs/N/stdout/` endpoint that's currently used for project updates, inventory updates, and system jobs)

Ad-hoc Command Support
---------------------------
Ad-hoc commands already record events from the `ansible` command via a callback plugin.  The UI does not use these events, however, so it must be updated to do so.

- [ ] modify the adhoc command details UI to display adhoc command event data instead of raw stdout

Project Update Support
------------------------
Project updates are implemented as an Ansible playbook, and should have first-class support for an event model, similar to Job Templates and Adhoc Commands:

- [x] introduce a new model, `ProjectUpdateEvent`, which follows the same pattern as JobEvent/AdhocCommandEvent
- [x] introduce a new API endpoint for listing/filtering `ProjectUpdateEvent` objects, likely something like `/api/v2/project_updates/N/events/`
- [x] add callback plugin support for capturing/recording project update events
- [x] add support for "downloading" raw stdout which is composed from event data (similar to how this works for Job and AdhocCommand events)

Event Support for Non-Ansible Jobs
------------------------------------
"Events" are a construct of Ansible runs within awx, and awx has several workloads that are not actually run via Ansible.  Currently, these are Inventory Updates, which are invoked as a Python binary, `awx-manage inventory_import`, and System Jobs, which are invoked as a Python binary, e.g., `awx-manage cleanup_jobs`.

- [x] introduce a new model for `StdoutEvent`.  This will service inventory updates, system jobs, and anything in the future that doesn't actually run through ansible and generate "real" events.
- [x] introduce new API endpoints for listing/filtering these objects, e.g., `/api/v2/inventory_updates/N/events/` and `/api/v2/system_jobs/N/events/`
- [x] add support for capturing stdout from these commands and generating (in awx) one event per line of output
- [x] add support for "downloading" raw stdout which is composed from event data (similar to how this works for Job and AdhocCommand events)

Additional Cleanup Tasks
-------------------------
- [x] To improved `main_unifiedjob` performance for inventory updates with large (10+MB) stdout, we added several ORM optimizations into 3.1/3.2, such as https://github.com/ansible/awx/commit/1bb6c17fe27d0deacc3bd0a3b763903fa35725df; when stdout is moved to a deprecated table and no longer part of `main_unifiedjob`, these optimizations and their associated code can go away.

Support for Legacy Stdout Content (Project Updates, Inventory Updates, System Jobs)
---------------------------------------------------------------------------------------
Open question: should we support "legacy stdout blobs" in the same per-line job event interface (cc: @wenottingham)?  Options include:
* Having code in the API serializer that detects legacy stdout content and creates per-line "virtual" events on the fly.  This could be slow for large stdout content.
* Not displaying legacy content via job events in the UI at all, but directing users to "download" the stdout raw content.

### Other Migration Notes
As noted in https://github.com/ansible/awx/issues/200#issuecomment-336227384, to support legacy stdout content, we'll likely need to migrate it to another deprecated table so that it no longer lives on the `main_unifiedjob` table.  We'll need to come up with a memory-efficient way to accomplish this in a migration that does _not_ involve shuffling around large strings in Python.

For "new" jobs, we'll stop storing raw stdout altogether.  This means we can remove quite a lot of code that:
* Stores raw stdout content on the local file system; this is subject to a variety of issues for clustered installations
* Stop storing raw stdout content as a raw blob in the database; when users download stdout, we recompose it every time via job events.
* The concept of "content length" UI display restrictions might go away; awx currently has a setting, `STDOUT_MAX_BYTES_DISPLAY`, which we use to determine when to truncate stdout for large output.  We'll need to consider what affect this will have (and the behavior we desire) for Inventory Updates that have many-MB output (and thus, lots of events) cc @wenottingham

Backward Compatibility Concerns
----------------------------------
We can continue to support the `/api/v2/jobs/N/stdout/` endpoint via event reconstruction across the board, but there are a number of places we've traditionally displayed stdout data in APIs that will become prohibitively expensive to do so.  For example:

`/api/v2/jobs/N/` currently contains a key, `result_stdout`, which we probably should remove in favor of explicit requests to `/api/v2/jobs/N/stdout/`  @wenottingham @matburt @AlanCoding how do you feel about this?

**Comment by AlanCoding** [comment]
> Having code in the API serializer that detects legacy stdout content and creates per-line "virtual" events on the fly. This could be slow for large stdout content.

What's the current size limit someone can get away with without pexpect crashing?

If the API was going to dynamically create "fake" job events, why not just save them to the database while you're at it? (i.e. migrate on-demand in the request-response cycle)

I would anticipate serious problems with pagination. This would require overriding the view at such an early stage in its rendering, that it would make me nervous. You would be inspecting the parent object qualities before generating the queryset, which I guess could work, but down-selecting for line-ranges won't be easy, that's why I think the on-demand migration might be the easiest.

**Comment by AlanCoding** [comment]
> `/api/v2/jobs/N/` currently contains a key, `result_stdout`, which probably should remove in favor of explicit requests to `/api/v2/jobs/N/stdout/` 

I think we should absolutely do this, although it will be a backward-incompatible break in pattern.

We should also consider what to do about result traceback. This might be smaller in size, but we may not be able to guarantee that. Perhaps there could be a marker on the new event model to denote that the text was coming through stderr.

Last comment: I'm very interested in the future of OutputEventFilter here. We could break this up into multiple classes for the different types of jobs, or we could hook into the logger itself. Like I was saying earlier, the logger statements give us a very free hand.

No wait, this is my last comment: With the inventory imports, I am frustrated that the `awx-manage` tasks have such a huge level of indirection. I could see us actually running system jobs inside of the celery context itself, just in a try-except block with the standard out handle pointing to where we want it to go. This isn't necessary, but I think it would be fully compatible with all the other things we want to do.

---
That wraps up my comments for this planning. I think you've done a great job @ryanpetrello and have a high degree of confidence in you carrying through this feature to its end. The history of ripping out `run_pexpect` already kind of exposed these rifts, and you know the shortcomings of our standard out wrapper very well.

**Comment by gconsidine** [comment]
## UI

Piggybacking on @ryanpetrello's comment, I'm including some UI-specific notes/questions here for discussion.

### Setup
Some initial setup in Leveraging the e2e infrastructure and the replay stdout utility to have easy access to a full spectrum of stdout (ad-hoc, inventory, project update, system jobs) at varying rates and sizes

### Managing data
Solve the problem of managing stdout events without attempting to solve the display issues at the same time. In other words, make sure the models/services in place can handle the following:

  - Receiving and ordering events
  - Discarding events once certain thresholds are met (some UX considerations needed here)
  - Fetching and merging in event data by line number (e.g. grab x lines before/after line y)
  - Managing searching/filtering
  - Downloading stdout

### Display

Once the model/data functionality is firmed up a bit, implement a sliding window approach to display the stdout. There are a few different paths we can take here depending on how the performance is looking after a bit of experimentation:

  - Display organized and styled output in its entirety (without dynamicfunctionality) and add dynamic functionality on the fly to lines in view asthe user scrolls up/down.
  - Display organized, styled, and dynamic output but only for as many lines that are displayed +/- some number of lines to allow for time to fetch more data as the user scrolls up/down.

### Optimization considerations and options

**Web Workers API**

Potentially useful on the model/data side for ordering events. Depending on how templating is approached, HTML snippets (containing span tags and so on to style, color output) could be generated here then handed off to be inserted into the DOM.

**Angular-only Approach**
Good news is performance-related gotchas are pretty well established for Angular 1.x since it's been out in the wild for a while now.

Be frugal as possible in use of watchers and aware of Angular functionality implicitly setting watchers. Each watcher set means a longer wait on every digest cycle. It can be compounded by watchers that watch watchers. 
  
Examples:
  - Use bind-once (as in `{{:: someValue }}`)
  - Avoid filters
  - Avoid ng-repeat
  - Use $watchCollection v. $watch if necessary at all

**Bypass Angular**

A complete bypass might be impractical. However, there might be steps somewhere between requesting/receiving data and rendering that'd benefit from the performance of plain old JavaScript.

### Questions

- How much of this can the API and UI team do in tandem? If the request/response structure can be loosely nailed down early on, that'd be enough to get started but leads into the next question...

- Will the stdout replay event utility be updated to match changes in the API? Or in the meantime, what's the level of effort involved in modifying it to spit back mock data based on how things currently look or will look? I'd be happy to help here if I can.

**Comment by jaredevantabor** [comment]
@gconsidine and I had an offline discussion about also consolidating the states for the stdout pages so that the stdout state is a child state of the `jobs` state. The purpose of this will be to maintain query params via the angular-breadcrumb library. related card: https://github.com/ansible/ansible-tower/issues/6455

**Comment by ruben-herold** [comment]
Hello, is there any workaround until its not resolved ?

can't see any progress or output from adhoc jobs

**Comment by ryanpetrello** [comment]
@ruben-herold there _is_ one workaround I know of, though it's still not ideal; there is a download icon in the stdout pane - if you click on that, it should download a text file containing the output.  On the next page load, the stdout *should* display.

**Comment by ruben-herold** [comment]
@ryanpetrello  sorry don't work here. If I download I get a textfile with:

        
{"detail":"Not found."}

Log shows:

2017-11-16 22:15:21,168 ERROR    django.request Internal Server Error: /api/v2/ad_hoc_commands/40/stdout/
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/generics.py", line 246, in dispatch
    return super(APIView, self).dispatch(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 466, in dispatch
    response = self.handle_exception(exc)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/views.py", line 463, in dispatch
    response = handler(request, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python2.7/site-packages/rest_framework/generics.py", line 210, in get
    return self.retrieve(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/awx/api/views.py", line 4353, in retrieve
    write_fd = open(unified_job.result_stdout_file, 'w')
IOError: [Errno 2] No such file or directory: u'/var/lib/awx/job_status/40-48f8ebd0-cb1b-11e7-841e-0242ac110006.out'

The file /var/lib/awx/job_status/40-48f8ebd0-cb1b-11e7-841e-0242ac110006.out  existst:


     docker container exec  7f3ca2869c9b ls -la  /var/lib/awx/job_status/40-48f8ebd0-cb1b-11e7-841e-0242ac110006.out 
-rw-r--r--. 1 root root 102207 Nov 16 22:12 /var/lib/awx/job_status/40-48f8ebd0-cb1b-11e7-841e-0242ac110006.out

**Comment by ruben-herold** [comment]
What I has missed last night was why:

write_fd = open(unified_job.result_stdout_file, 'w')
IOError: [Errno 2] No such file or directory: u'/var/lib/awx/job_status/40-48f8ebd0-cb1b-11e7-841e-0242ac110006.out'

do you try to open writeable if it is a read request?

**Comment by ryanpetrello** [comment]
@ruben-herold what you're seeing is the old mechanism for caching stdout data on the local disk (which is what we'll be replacing with the feature described in this PR, because as you've seen it's unreliable).  A big reason for the work described in this issue relates to the fact that this caching doesn't work in clustered environments (because every awx node has a different disk).  The workaround I suggested is a sneaky hack to work around that limitation to temporarily generate the cached stdout on the local node you care about.

Normally the line you're seeing that exception on will work just fine (despite it writing a file in a `GET` request, which is pretty gross).  For some reason, though, it's blowing up for you, and I suspect it's related to the container environment.  @matburt the exception we're seeing here looks *suspiciously* like the one we saw at:

https://github.com/ansible/awx/issues/492

**Comment by ryanpetrello** [comment]
@ruben-herold out of curiosity, if you change this line:

```python
write_fd = open(unified_job.result_stdout_file, 'w')
```

...to this:

```python
import os
write_fd = os.fdopen(unified_job.result_stdout_file, 'w')
```

...does the download button start working?

**Comment by khobbits** [comment]
Just thought I'd drop this here, in a clear way, as a potential work around for people running the currently published docker containers:

If you get the "stdout capture is missing" as the result of an ad-hoc job run, you should be able to click the download button, and get forwarded to a URL like:
http://awx/api/v2/ad_hoc_commands/18/stdout/?format=txt_download

This will likely give you the output: { "detail": "Not found." }

However, if you rewrite that url, to be:
http://awx/api/v2/ad_hoc_commands/18/events/

(dropping the /stdout and replacing it with events/)

You will get a full output, which does show the output of the stdout, although you'd have to search for it, you will see one result per host...

For people more json minded, it's located in: results.[].event_data.res.stdout

**Comment by ryanpetrello** [comment]
Thanks @khobbits, that's another option.  In fact, the download button I suggested is _actually_ doing the same thing under the hood (parsing the stdout components from the event JSON: https://github.com/ansible/awx/blob/c723ba52893b136eb210dc175af317d4a4e3d484/awx/api/views.py#L4436

It just seems like it's broken due to the bug that @ruben-herold is currently encountering.

**Comment by dw-ec** [comment]
Hypothetically†, as a kludge of a work-around, if I were to mount a docker volume at /var/lib/awx/job_status on awx_task, and then mount the same volume read-only at the same path on awx_web, what's the worst that could happen? 

†Because I've just done that, and I can now see the stdout of ad_hoc & other jobs from the command line and in the web interface, which I couldn't before:

```
$ tower-cli ad_hoc launch -i myinv --credential mykey --module-args 'echo Hello stdout!'
Resource changed.
== =========================== ====== ======= ========= =========== 
id           created           status elapsed inventory module_name 
== =========================== ====== ======= ========= =========== 
42 2017-11-30T20:52:46.940080Z new    0.0             4 command
== =========================== ====== ======= ========= =========== 

$ tower-cli ad_hoc stdout 42
Identity added: /tmp/awx_42_xXxXxX/credential_2 (/tmp/awx_42_xXxXxX/credential_2)
myhost | SUCCESS | rc=0 >>
Hello stdout!


OK. (changed: false)
```

Same question but about those _manual SCM type_ projects imported into /var/lib/awx/projects that also aren't being synched between awx_host and awx_web; and for that matter any other bit of var data that's been scuppered by containerisation?

**Comment by ryanpetrello** [comment]
@id-ec 
> what's the worst that could happen?

You could be ashamed of yourself 😉 Points for ingenuity 😄 

Seriously, though, there are definitely creative ways to just "make this work" for now, and what you've mentioned _would_ technically work with a _single_ `awx_task` container and an explicit volume mount.  But as you said, it's definitely a kludge.  In the future, `awx` is moving towards a model where this content just doesn't live on disk anymore (this ticket/issue).

As for manual projects, the semi-official recommendation for clustered / containerized awx installs is "use source control".

**Comment by aizuddin85** [comment]
@ryanpetrello  I have the same bugs encountered by @ruben-herold.

using `os.fdopen` still not fix anything for me. I notice that awx_web container does not see what is in the awx_task job_status folder, hence,  as the workaround, I did volume bind on both awx_task and awx_web container to the same shared NFS. 

As I`m using local_docker, I did intrusive reconfiguration on installer/local_docker/tasks/main.yaml, adding volumes:

awx_task
```
-- truncated --
volumes:
      - /awx_shared:/var/lib/awx/job_status:rw
-- truncated --
```

awx_web
```
-- truncated --
volumes:
      - /awx_shared:/var/lib/awx/job_status:ro
-- truncated --
```  
df
```
172.28.0.1:/awx_shared_qtree    /awx_shared
```
With above workaround, the download is working again. Looking forward for this PR to be completed then.

**Comment by ryanpetrello** [comment]
@aizuddin85 thanks for the reply.  Glad to hear you've got a workaround for now - we're working on accomplishing this in a better way in the near future!

**Comment by ryanpetrello** [comment]
@jaredevantabor @gconsidine API changes for this are merged and should have everything necessary for UI work.

**Comment by Glutexo** [comment]
Just wondering, is there now a way (some workaround until this gets fixed) to get `stdout` of a `project_update`, which now has only `stdout capture is missing`? We are encountering this quite often.

**Comment by AlanCoding** [comment]
That sounds like https://github.com/ansible/awx/issues/1687. The UI is planned to switch to consuming event-based project update output, but clients using the old way of retrieving stdout still work, so there's no imminent usability problem from this.

**Comment by Glutexo** [comment]
@AlanCoding I guess this is not a UI problem. I’m seeing this not only in Tower UI on a `/scm_update/#{id}` page as `stdout capture is missing`, but also in the API:

* `/api/v1/project_updates/#{id}/stdout/` returns `stdout capture is missing` too
* `/api/v1/project_updates/#{id}/stdout/?format=txt_download` returns `{'error': "Error generating stdout download file: [Errno 2] No such file or directory: u''"}`

**Comment by AlanCoding** [comment]
interesting, I cannot reproduce that.

**Comment by ryanpetrello** [comment]
> Just wondering, is there now a way (some workaround until this gets fixed)

@Glutexo it is fixed; what version of awx are you running?  If you're still seeing `stdout capture is missing` for a project update, you're likely running an older version of awx and need to upgrade to resolve this issue.

**Comment by matburt** [comment]
Is this issue still pending UI changes?

**Comment by ryanpetrello** [comment]
@matburt yes, I believe so.  I think @jakemcdermott is still working on it?  Specifically, the views for Inventory Updates, Adhoc, and System Jobs need to be changed to use the new event model (instead of stdout download polling).

It currently _works_ in a backwards compatible way (because the `/stdout/` endpoint _composes_ the content on-the-fly via job event data), but we want to actually update these interfaces to use real events and websocket notification streaming.

**Comment by jakemcdermott** [comment]
> Is this issue still pending UI changes?

Yes.

**Comment by Glutexo** [comment]
@ryanpetrello I am encountering this with Ansible Tower 3.0.1, which is kind of older by now. I’ll come back in case I reproduce this issue with a newer version.

**Comment by Glutexo** [comment]
@ryanpetrello Still – is there some workaround for the older versions? Like some property that could be received from the API, either for the update job or for the whole project, that would point to the fact that the stdout won’t be captured? This would allow us to postpone the update a little bit in favor of being sure to have the stdout capture.

**Comment by ryanpetrello** [comment]
@Glutexo this comment thread is for the open source awx project, and Tower 3.0.1 is _quite_ old (in fact, it's EOL and no longer supported by Red Hat as of last month).  Could you reach out to me or @ghjm privately so we can discuss this further?

**Comment by Glutexo** [comment]
@ryanpetrello I just thought that this issue has the same cause both in Tower and AWX. But as we talked in the BZ, it does not happen in newer versions. Furthermore, the issue happens because of not so smart project update triggering. I believe that fixing this on our side would make this issue disappear even for the legacy Tower versions. Thanks for your help! I’ll reach to you personally in case I’d need some more assistance.

**Comment by jakemcdermott** [comment]
We've addressed this issue (and others!) downstream so we've closed this issue as resolved - hang tight and the changes will be arriving in AWX in the future.

**Comment by llbbl** [comment]
Is this fixed in 1.0.6 ? Looks like the comment from Jake was after the tag was added. Would be worth upgrading just for this.
